upcloud/resource_upcloud_objectstorage.go

```diff
@@ -380,3 +438,117 @@ func modifyObjectStorage(client *service.Service, req *request.ModifyObjectStora
+	configVal := d.Get("access_key").(string)
+
+	// If config value is set to something else then empty string, just use it
+	if configVal != "" {
+		return configVal, false, nil
+	}
+
+	// If config value is empty string, use environment variable
+	objectStorageName := d.Get("name").(string)
+	envVarKey := generateObjectStorageEnvVarKey(AccessKeyEnvVarPrefix, objectStorageName)
+	envVarValue, envVarSet := os.LookupEnv(envVarKey)
+
+	if !envVarSet {
+		return "", false, fmt.Errorf("access_key config field for object storage %s is set to empty string and environment variable %s is not set", objectStorageName, envVarKey)
+	}
+
+	length := len(envVarValue)
+
+	if length < accessKeyMinLength {
+		return "", false, fmt.Errorf("access_key set in environment variable %s is too short; minimum length is %d, got %d", envVarKey, accessKeyMinLength, length)
+	}
+
+	if length > accessKeyMaxLength {
+		return "", false, fmt.Errorf("access_key set in environment variable %s is too long; maximum length is %d, got %d", envVarKey, accessKeyMaxLength, length)
+	}
+
+	return envVarValue, true, nil
+}
+
+// Attempts to get secret key.
+// Second return value is a bool, set to true if key value was revtrived from env variable
+func getSecretKey(d *schema.ResourceData) (string, bool, error) {
+	configVal := d.Get("secret_key").(string)
+
+	// If config value is set to something else then empty string, just use it
+	if configVal != "" {
+		return configVal, false, nil
+	}
+
+	// If config value is empty string, use environment variable
+	objectStorageName := d.Get("name").(string)
+	envVarKey := generateObjectStorageEnvVarKey(SecretKeyEnvVarPrefix, objectStorageName)
+	envVarValue, envVarSet := os.LookupEnv(envVarKey)
+
+	if !envVarSet {
+		return "", false, fmt.Errorf("secret_key config field for object storage %s is set to empty string and environment variable %s is not set", objectStorageName, envVarKey)
+	}
+
+	length := len(envVarValue)
+
+	if length < secretKeyMinLength {
+		return "", false, fmt.Errorf("secret_key set in environment variable %s is too short; minimum length is %d, got %d", envVarKey, secretKeyMinLength, length)
+	}
+
+	if length > secretKeyMaxLength {
+		return "", false, fmt.Errorf("secret_key set in environment variable %s is too long; maximum length is %d, got %d", envVarKey, secretKeyMaxLength, length)
+	}
+
+	return envVarValue, true, nil
+}
+
+func generateObjectStorageEnvVarKey(prefix, objectStorageName string) string {
+	name := strings.ToUpper(strings.Replace(objectStorageName, "-", "_", -1))
+	return fmt.Sprintf("%s%s", prefix, name)
+}
+
+type objectStorageKeyType string
+
+const (
+	objectStorageKeyTypeAccess objectStorageKeyType = "access_key"
+	objectStorageKeyTypeSecret objectStorageKeyType = "secret_key"
+)
+
+func createKeyValidationFunc(attrName objectStorageKeyType, minLength, maxLength int) schema.SchemaValidateDiagFunc {
+	return func(val interface{}, path cty.Path) diag.Diagnostics {
+		key := val.(string)
```

Comment:
This is not big one but we could check if type assertion fails:
```
key, ok := val.(string)
if !ok {
   return diag.FromErr(fmt.Errorf("expected type of %s to be string", val))
}
```

Reply:
It technically should not be possible, since the type is defined in schema. Then again the runtime type assertion errors do happen from time to time in development, so I guess it would be good to get into habit of checking those, even if terraform sdk guarantees them :P